[CodeGen][X86] Fix lowering of tailcalls when `-ms-hotpatch` is used

aganea opened the pull request:
Previously, tail jump pseudo-opcodes were skipped by the `encodeInstruction()` call inside `X86AsmPrinter::LowerPATCHABLE_OP`. This caused emission of a 2-byte NOP and dropping of the tail jump.

With this PR, we change `PATCHABLE_OP` to not wrap the first `MachineInstr` anymore, but inserting itself before the instruction, leaving the instruction unaltered. At lowering time in `X86AsmPrinter`, we now "look ahead" for the next non-pseudo `MachineInstr` and lower+encode it, to inspect its size. If the size is below what `PATCHABLE_OP` expects, it inserts NOPs; otherwise it does nothing. That way, now the first `MachineInstr` is always lowered as usual even if `"patchable-function"="prologue-short-redirect"` is used.

Fixes https://github.com/llvm/llvm-project/issues/76879, https://github.com/llvm/llvm-project/issues/76958 and https://github.com/llvm/llvm-project/issues/59039

aganea committed:
[CodeGen][X86] Fix lowering of tailcalls when `-ms-hotpatch` is used

Previously, tail jump pseudo-opcodes were skipped by the `encodeInstruction()` call inside `X86AsmPrinter::LowerPATCHABLE_OP`. This caused emission of a 2-byte NOP and dropping of the tail jump

Also, the `PatchableFunction` pass didn't properly update the call site info for the existing MachineInstr, after wrapping it into a `TargetOpcode::PATCHABLE_OP`.

KanRobert commented on llvm/test/CodeGen/X86/patchable-prologue-tailcall.ll:
```diff
@@ -0,0 +1,69 @@
+;     return __builtin_malloc(count);
+; }
+
+; ModuleID = 'patchable-prologue-tailcall.cpp'
+source_filename = "patchable-prologue-tailcall.cpp"
```
Delete?

KanRobert commented on llvm/test/CodeGen/X86/patchable-prologue-tailcall.ll:
```diff
@@ -0,0 +1,69 @@
+; Function Attrs: mustprogress nofree nounwind willreturn allockind("alloc,uninitialized") allocsize(0) memory(inaccessiblemem: readwrite)
+declare dso_local noalias noundef ptr @malloc(i64 noundef) local_unnamed_addr #3
+
+attributes #0 = { mustprogress nounwind sspstrong uwtable "min-legal-vector-width"="0" "no-trapping-math"="true" "patchable-function"="prologue-short-redirect" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cmov,+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
+attributes #1 = { "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cmov,+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
+attributes #2 = { mustprogress nofree nounwind sspstrong willreturn memory(inaccessiblemem: readwrite) uwtable "min-legal-vector-width"="0" "no-trapping-math"="true" "patchable-function"="prologue-short-redirect" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cmov,+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
+attributes #3 = { mustprogress nofree nounwind willreturn allockind("alloc,uninitialized") allocsize(0) memory(inaccessiblemem: readwrite) "alloc-family"="malloc" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cmov,+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
```
Keep the attribute `"patchable-function"="prologue-short-redirect"` only?

KanRobert commented on llvm/test/CodeGen/X86/patchable-prologue-tailcall.ll:
```diff
@@ -0,0 +1,69 @@
+!3 = !{i32 8, !"PIC Level", i32 2}
+!4 = !{i32 7, !"uwtable", i32 2}
+!5 = !{i32 1, !"MaxTLSAlign", i32 65536}
+!6 = !{!"clang version 18.0.0git (https://github.com/llvm/llvm-project.git 0ccf9e29e9626a3a6813b35608c8959178c50a6f)"}
```
Remove !0, !1, !6?

KanRobert commented on llvm/lib/Target/X86/X86MCInstLower.cpp:
```diff
@@ -994,8 +995,11 @@ void X86AsmPrinter::LowerPATCHABLE_OP(const MachineInstr &MI,
-  if (!EmptyInst)
+  if (!EmptyInst) {
+    if (Opcode != convertTailJumpOpcode(Opcode))
+      OutStreamer->AddComment("TAILCALL");
```
This line is not tested at all?

KanRobert commented on llvm/lib/CodeGen/PatchableFunction.cpp:
```diff
@@ -87,6 +87,9 @@ bool PatchableFunction::runOnMachineFunction(MachineFunction &MF) {
     MIB.add(MO);
 
+  if (FirstActualI->shouldUpdateCallSiteInfo())
+    MF.eraseCallSiteInfo(&*FirstActualI);
```
Maybe we don't need this. According to the description above, `PatchableFunction` needs the first instruction is two-byte long.

http://xxeo.com/single-byte-or-small-x86-opcodes

One-byte instruction is rare in X86. `TailCall` can be transformed to either  a `call` or a `jmp`, which is always longer than 2 bytes. I think we can just skip emitting patchable-op if the first instruction is a `TailCall`.

sylvain-audi commented:
I remember trying a similar approach: Making the pass skip the functions for which the first instruction's `isCall()` returns true, assuming that all calls/jumps are >= 2 bytes long.

One one hand, I think it fixed the issue for the repro provided in the 2 issues, so it could be considered a workaround (and it is not worse than before this code change).

However I don't think it should close https://github.com/llvm/llvm-project/issues/59039, because of the remaining design issues:

1) Wrapping an instruction into a `PATCHABLE_OP` loses information that would be needed at lowering time. 
In `X86AsmPrinter::emitInstruction`, the wrapped opcode doesn't go through `X86AsmPrinter::emitInstruction` itself, which  I'm pretty sure can become problematic. The main concern here is that `-fms-hotpatch` could alter the expected first instruction (instead of only guaranteeing its minimum size).

 2) `PATCHABLE_OP` takes the size as parameter (in reality it is always 2), but it is hard for `PatchableFunction` pass to know precisely the exact size of the instruction at that point. Especially, with your change, we start to assume that the size is 2. Maybe it should be changed to only handle a size of 2, and simplify the lowering code.


What do you think?

KanRobert commented:
@sylvain-audi
For 1, I think we can avoid introducing the wrapper `PATCHABLE_OP` and instead add `MIFlag` (/llvm/include/llvm/CodeGen/MachineInstr.h) for your purpose.

For 2, instruction length is only determined before emitting. I think it's impossible to know that in `PatchableFunction`. if the size is always 2, we should not overdesign.

aganea committed:
Change how `"patchable-function"="prologue-short-redirect"`works.

Instead of `PATCHABLE_OP` wrapping the first `MachineInstr`, it now inserts itself before it, leaving the instruction unaltered. At lowering time in `X86AsmPrinter`, we now "look ahead" for the next non-pseudo `MachineInstr` and lower+encode it, to inspect its size. If the size is below what `PATCHABLE_OP` expects, it inserts NOPs; otherwise it does nothing

aganea commented:
So I did something different to address the points raised. Instead of `PATCHABLE_OP` wrapping the first `MachineInstr`, it now inserts itself before it, leaving the instruction unaltered. At lowering time in `X86AsmPrinter`, we now "look ahead" for the next non-pseudo `MachineInstr` and lower+encode it, to inspect its size. If the size is below what `PATCHABLE_OP` expects, it inserts NOPs; otherwise it does nothing. I think this should address @sylvain-audi 's concerns. PTAL.

If we wanted in a future PR we could change the attribute `"patchable-function"="prologue-short-redirect"` to `"patchable-function"="2"`. That would make `PATCHABLE_OP` more flexible, and the X86 backend already supports large sizez NOPs (up to 15, if the target CPU supports it).

Also, I've removed the `X86::PUSH64r` optimization which cannot be applied anymore in this situation. Thus the change in the `patchable-prologue.ll` test. +@sanjoy

aganea committed:
Documentation.

sylvain-audi commented on llvm/lib/Target/X86/X86MCInstLower.cpp:
```diff
@@ -948,24 +948,26 @@ void X86AsmPrinter::LowerASAN_CHECK_MEMACCESS(const MachineInstr &MI) {
+  unsigned MinSize = MI.getOperand(0).getImm();
+
+  if (NextMI) {
+    // Lower the next MachineInstr to find its byte size.
```
Would it be possible here to use `X86AsmPrinter`'s `emitInstruction` to lower `NextMI`?
That way it should guarantee we match what will be generated when handling the next `MachineInstr`.
But I don't know how feasible it is, as it may have some edge effects...

aganea replied:
I fear that would require a massive refactoring. `X86AsmPrinter::emitInstruction` is manipulating state (like `OutStreamer` or the counting) that we would need to reset if the instruction isn't large enough (for patchable op).
I used `X86MCInstLower::Lower` which should bring us closer to what you suggested. I can remove the reference to https://github.com/llvm/llvm-project/issues/59039 is you think more is needed in the future.

sylvain-audi replied:
To my mind you can keep the reference to #59039 , since your refactor makes this system quite more robust.

sylvain-audi commented:
> Also, I've removed the `X86::PUSH64r` optimization which cannot be applied anymore in this situation. Thus the change in the `patchable-prologue.ll` test. +@sanjoy

To keep the optimization, is there a (clean) way we could "tell" the `X86AsmPrinter` to ignore the emission of the next instruction?

aganea commented:
> > Also, I've removed the `X86::PUSH64r` optimization which cannot be applied anymore in this situation. Thus the change in the `patchable-prologue.ll` test. +@sanjoy
> 
> To keep the optimization, is there a (clean) way we could "tell" the `X86AsmPrinter` to ignore the emission of the next instruction?

I was under the impression that was initially a low hanging fruit optimization, but I'm not sure it's required ( @sanjoy ). I'll look into what you suggest if that's required.

aganea commented:
Another thing that @sylvain-audi raised in a offline discussion, is that `PatchableFunction` pass adds a 16-byte alignement for every single function in the PE, which increased the size and is not required stricto sensu. It is only the usage of `/functionpadmin` linker option that does the extra padding at the begining of the function and could/should take this alignement decision.

tycho commented:
I'm the original reporter to @MolecularMatters about the issue, and I've been running with the Live++ build flags in my project for the past 12 hours with the changes in this PR applied on top of 8fb8ad66c95a51b82e5c2876ed925b5512ce6b83 and it seems to be fine now without needing `-fno-optimize-sibling-calls`

MolecularMatters commented:
@aganea @tycho @sylvain-audi If somebody could make a clang/lld build with the patch available for me, I'd be happy to run this against the Live++ test suite.

sylvain-audi commented:
> @aganea @tycho @sylvain-audi If somebody could make a clang/lld build with the patch available for me, I'd be happy to run this against the Live++ test suite.

@MolecularMatters I sent you a build with this PR modification, based on 17.0.6.
Just mentioning it so @tycho  and @aganea don't need to worry about preparing/sending it.

aganea commented:
@MolecularMatters Did you have a chance to test this PR? I’d like to merge it before the branch cutoff. Thanks!

MolecularMatters commented:
@aganea Thanks for reminding me!
Yes, I just tested this, seems to work fine.

aganea commented:
> @aganea Thanks for reminding me! Yes, I just tested this, seems to work fine.

Thanks!

To all other reviewers: is this good to go? Can anyone approve please?

aganea commented:
Ping!

KanRobert commented on llvm/lib/Target/X86/X86MCInstLower.cpp:
```diff
@@ -948,24 +948,26 @@ void X86AsmPrinter::LowerASAN_CHECK_MEMACCESS(const MachineInstr &MI) {
-  for (auto &MO : drop_begin(MI.operands(), 2))
-    if (auto MaybeOperand = MCIL.LowerMachineOperand(&MI, MO))
-      MCI.addOperand(*MaybeOperand);
+  // Find the next MachineInstr in this MBB.
+  const MachineInstr *NextMI = MI.getNextNode();
+  while (NextMI) {
+    if (!NextMI->isMetaInstruction())
+      break;
+    NextMI = NextMI->getNextNode();
+  }
```
```suggestion
  // Find the next MachineInstr in this MBB.
 auto NextMI = std::find_if(MI.getIterator(), MI.getParent()->end(), [](auto &II) { return !II.isMetaInstruction(); });

```

KanRobert commented on llvm/lib/Target/X86/X86MCInstLower.cpp:
```diff
@@ -948,24 +948,26 @@ void X86AsmPrinter::LowerASAN_CHECK_MEMACCESS(const MachineInstr &MI) {
-  for (auto &MO : drop_begin(MI.operands(), 2))
-    if (auto MaybeOperand = MCIL.LowerMachineOperand(&MI, MO))
-      MCI.addOperand(*MaybeOperand);
+  // Find the next MachineInstr in this MBB.
```
Drop this comment, it's self-explained if we use `find_if`

KanRobert commented on llvm/test/CodeGen/X86/patchable-prologue-tailcall.ll:
```diff
@@ -0,0 +1,55 @@
+;     return __builtin_malloc(count);
+; }
+
+target datalayout = "e-m:w-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-pc-windows-msvc19.38.33133"
```
Remove these and add `-mtriple=x86_64-windows-msvc` to the command line

KanRobert commented on llvm/test/CodeGen/X86/patchable-prologue-tailcall.ll:
```diff
@@ -0,0 +1,55 @@
+target datalayout = "e-m:w-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-pc-windows-msvc19.38.33133"
+
+define dso_local noundef ptr @"?mi_new_test@@YAPEAX_K@Z"(i64 noundef %count) local_unnamed_addr "patchable-function"="prologue-short-redirect" {
+entry:
+  %call = tail call ptr @mi_new(i64 noundef %count)
+  ret ptr %call
+}
+
+declare dso_local ptr @mi_new(i64 noundef) local_unnamed_addr
+
+define dso_local noalias noundef ptr @"?builtin_malloc_test@@YAPEAX_K@Z"(i64 noundef %count) local_unnamed_addr "patchable-function"="prologue-short-redirect" {
+entry:
+  %call = tail call ptr @malloc(i64 noundef %count)
+  ret ptr %call
+}
```
Remove dso_local

KanRobert commented on llvm/test/CodeGen/X86/patchable-prologue-tailcall.ll:
```diff
@@ -0,0 +1,55 @@
+target datalayout = "e-m:w-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-pc-windows-msvc19.38.33133"
+
+define dso_local noundef ptr @"?mi_new_test@@YAPEAX_K@Z"(i64 noundef %count) local_unnamed_addr "patchable-function"="prologue-short-redirect" {
```
Remove local_unnamed_addr

KanRobert replied:
Rename the function as we did in llvm/test/CodeGen/X86/patchable-prologue.ll. So it would not give readers the illusion that this function name may affect functionality.

KanRobert commented on llvm/test/CodeGen/X86/patchable-prologue-tailcall.ll:
```diff
@@ -0,0 +1,55 @@
+; CHECK-NEXT: jmp     malloc                          # TAILCALL
+
+; // Built with: clang-cl.exe /c patchable-prologue-tailcall.cpp /O2 /hotpatch -Xclang -emit-llvm
+;
+; typedef unsigned long long size_t;
+; 
+; extern "C" {
+;     void* mi_new(size_t size);
+;  }
+;
+; void *mi_new_test(size_t count)
+; {
+;     return mi_new(count);
+; }
+; 
+; void *builtin_malloc_test(size_t count)
+; {
+;     return __builtin_malloc(count);
+; }
+
```
Drop this comment, it is just saying how to write a tailcall, which should be straightforward for developers.

KanRobert reviewed (approved):
LGTM with some suggestion

aganea committed:
Simplify test and change code as suggested

aganea commented:
As suggested

sylvain-audi commented:
> @sylvain-audi Would you prefer that we leave #59039 open?

No, you can close it in my opinion, thanks!
